File: scripts/bump-version.py
#!/usr/bin/env python3
import sys
import os
import re
import subprocess

def run_command(cmd, shell=False):
    """Helper to run command and exit if it fails."""
    print(f"Running: {' '.join(cmd) if isinstance(cmd, list) else cmd}")
    res = subprocess.run(cmd, shell=shell)
    if res.returncode != 0:
        print(f"Error: Command failed with exit code {res.returncode}", file=sys.stderr)
        sys.exit(res.returncode)

def main():
    # 1. Run checks first before doing anything
    print("=== Running pre-bump checks ===")
    run_command(["cargo", "fmt", "--all", "--", "--check"])
    run_command(["cargo", "clippy", "--all-targets", "--all-features", "--", "-D", "warnings"])
    run_command(["cargo", "test", "--all-targets", "--all-features"])
    print("=== Pre-bump checks passed successfully! ===\n")

    # 2. Read Cargo.toml
    cargo_toml_path = "Cargo.toml"
    if not os.path.exists(cargo_toml_path):
        print(f"Error: {cargo_toml_path} not found in current directory.", file=sys.stderr)
        sys.exit(1)

    with open(cargo_toml_path, "r") as f:
        content = f.read()

    # Find current version under [package]
    # We look for the first version field which is typically the package version
    match = re.search(r'^version\s*=\s*"([^"]+)"', content, re.MULTILINE)
    if not match:
        print("Error: Could not parse version from Cargo.toml", file=sys.stderr)
        sys.exit(1)

    current_version = match.group(1)
    print(f"Current package version: {current_version}")

    # Determine target version
    target_version = None
    if len(sys.argv) > 1 and sys.argv[1].strip():
        target_version = sys.argv[1].strip()
        print(f"Bumping to user-specified version: {target_version}")
    else:
        # Auto bump patch version (e.g. 0.1.0 -> 0.1.1)
        version_parts = current_version.split(".")
        if len(version_parts) != 3:
            print(f"Error: Version '{current_version}' is not in standard X.Y.Z semver format.", file=sys.stderr)
            sys.exit(1)
        try:
            patch = int(version_parts[2])
            version_parts[2] = str(patch + 1)
            target_version = ".".join(version_parts)
            print(f"Auto-bumping patch version: {current_version} -> {target_version}")
        except ValueError:
            print(f"Error: Patch version part '{version_parts[2]}' in '{current_version}' is not an integer.", file=sys.stderr)
            sys.exit(1)

    # 3. Update Cargo.toml
    # We only replace the first match of version="..." to avoid modifying dependency versions
    new_content, count = re.subn(
        r'^version\s*=\s*"[^"]+"',
        f'version = "{target_version}"',
        content,
        count=1,
        flags=re.MULTILINE
    )
    if count == 0:
        print("Error: Failed to replace version in Cargo.toml", file=sys.stderr)
        sys.exit(1)

    with open(cargo_toml_path, "w") as f:
        f.write(new_content)
    print("Updated Cargo.toml successfully.")

    # 4. Update Cargo.lock by running cargo check
    print("Running 'cargo check' to update Cargo.lock...")
    run_command(["cargo", "check"])

    # 5. Git commit and tag
    print("Staging version changes...")
    run_command(["git", "add", "Cargo.toml", "Cargo.lock"])

    commit_msg = f"chore: bump version to {target_version}"
    print(f"Committing changes (excluding git hooks with --no-verify)...")
    run_command(["git", "commit", "-m", commit_msg, "--no-verify"])

    tag_name = f"v{target_version}"
    print(f"Creating git tag: {tag_name}...")
    run_command(["git", "tag", "-a", tag_name, "-m", f"Release {tag_name}"])

    print("\n=======================================================")
    print("SUCCESS: Version bumped, committed, and tagged locally!")
    print(f"New Version: {target_version}")
    print(f"Git Tag:     {tag_name}")
    print("=======================================================")
    print("To push the changes and the tag to GitHub, run:")
    print(f"  git push origin && git push origin {tag_name} --no-verify")
    print("=======================================================")

if __name__ == "__main__":
    main()

File: scripts/bump-version.rs
use std::env;
use std::fs;
use std::process::{Command, exit};

fn run_command(cmd: &str, args: &[&str]) {
    println!("Running: {} {}", cmd, args.join(" "));
    let status = Command::new(cmd)
        .args(args)
        .status()
        .unwrap_or_else(|err| {
            eprintln!("Failed to execute command '{}': {}", cmd, err);
            exit(1);
        });
    if !status.success() {
        eprintln!("Error: Command '{}' failed with exit code {:?}", cmd, status.code());
        exit(status.code().unwrap_or(1));
    }
}

fn main() {
    // 1. Run checks first before doing anything
    println!("=== Running pre-bump checks ===");
    run_command("cargo", &["fmt", "--all", "--", "--check"]);
    run_command("cargo", &["clippy", "--all-targets", "--all-features", "--", "-D", "warnings"]);
    run_command("cargo", &["test", "--all-targets", "--all-features"]);
    println!("=== Pre-bump checks passed successfully! ===\n");

    // 2. Read Cargo.toml
    let cargo_toml_path = "Cargo.toml";
    let content = fs::read_to_string(cargo_toml_path).unwrap_or_else(|err| {
        eprintln!("Error reading {}: {}", cargo_toml_path, err);
        exit(1);
    });

    // Parse the current version. We find the version line within [package].
    let mut in_package = false;
    let mut current_version = None;
    let mut version_line_index = None;
    let lines: Vec<&str> = content.lines().collect();

    for (i, line) in lines.iter().enumerate() {
        let trimmed = line.trim();
        if trimmed == "[package]" {
            in_package = true;
            continue;
        }
        if in_package && trimmed.starts_with('[') {
            // Entered another section without finding version
            in_package = false;
        }
        if in_package && trimmed.starts_with("version") {
            // Parse version = "X.Y.Z"
            if let Some(start) = line.find('"') {
                if let Some(end) = line[start + 1..].find('"') {
                    let version = &line[start + 1..start + 1 + end];
                    current_version = Some(version.to_string());
                    version_line_index = Some(i);
                    break;
                }
            }
        }
    }

    let current_version = current_version.unwrap_or_else(|| {
        eprintln!("Error: Could not find version in [package] section of Cargo.toml");
        exit(1);
    });
    let version_line_idx = version_line_index.unwrap();

    println!("Current package version: {}", current_version);

    // Determine target version
    let args: Vec<String> = env::args().collect();
    let target_version = if args.len() > 1 && !args[1].trim().is_empty() {
        let val = args[1].trim().to_string();
        println!("Bumping to user-specified version: {}", val);
        val
    } else {
        // Auto bump patch version
        let parts: Vec<&str> = current_version.split('.').collect();
        if parts.len() != 3 {
            eprintln!("Error: Current version '{}' is not in standard X.Y.Z semver format.", current_version);
            exit(1);
        }
        let patch: u32 = parts[2].parse().unwrap_or_else(|_| {
            eprintln!("Error: Patch version part '{}' in '{}' is not an integer.", parts[2], current_version);
            exit(1);
        });
        let new_patch = patch + 1;
        let val = format!("{}.{}.{}", parts[0], parts[1], new_patch);
        println!("Auto-bumping patch version: {} -> {}", current_version, val);
        val
    };

    // 3. Update Cargo.toml content
    let mut new_lines = lines.clone();
    let old_line = lines[version_line_idx];
    let new_line = if let Some(start) = old_line.find('"') {
        if let Some(end) = old_line[start + 1..].find('"') {
            let prefix = &old_line[..start + 1];
            let suffix = &old_line[start + 1 + end..];
            format!("{}{}{}", prefix, target_version, suffix)
        } else {
            format!("version = \"{}\"", target_version)
        }
    } else {
        format!("version = \"{}\"", target_version)
    };

    new_lines[version_line_idx] = &new_line;
    let new_content = new_lines.join("\n") + "\n";

    fs::write(cargo_toml_path, new_content).unwrap_or_else(|err| {
        eprintln!("Error writing {}: {}", cargo_toml_path, err);
        exit(1);
    });
    println!("Updated Cargo.toml successfully.");

    // 4. Update Cargo.lock by running cargo check
    println!("Running 'cargo check' to update Cargo.lock...");
    run_command("cargo", &["check"]);

    // 5. Git commit and tag
    println!("Staging version changes...");
    run_command("git", &["add", "Cargo.toml", "Cargo.lock"]);

    let commit_msg = format!("chore: bump version to {}", target_version);
    println!("Committing changes (excluding git hooks with --no-verify)...");
    run_command("git", &["commit", "-m", &commit_msg, "--no-verify"]);

    let tag_name = format!("v{}", target_version);
    println!("Creating git tag: {}...", tag_name);
    run_command("git", &["tag", "-a", &tag_name, "-m", &format!("Release {}", tag_name)]);

    println!("\n=======================================================");
    println!("SUCCESS: Version bumped, committed, and tagged locally!");
    println!("New Version: {}", target_version);
    println!("Git Tag:     {}", tag_name);
    println!("=======================================================");
    println!("To push the changes and the tag to GitHub, run:");
    println!("  git push origin && git push origin {} --no-verify", tag_name);
    println!("=======================================================");
}
